src/FakeClient/GraphicsCaptureEventProcessor.h

```diff
@@ -0,0 +1,229 @@
+    while (currentBucket < distribution.size() &&
+           // checks if runningCount is less than centile/numCentiles of the
+           // population size, but keeps the calculation within the integral
+           // domain by scaling with numCentiles
+           (runningCount * numCentiles) < (populationSize * centile)) {
+      runningCount += distribution[currentBucket];
+      ++currentBucket;
+    }
+    operation(currentBucket);
+  }
+}
+}  // namespace
+
+namespace orbit_fake_client {
+
+// This implementation of CaptureEventProcessor mostly discard all events it receives, except for
+// keeping track of the calls to the frame boundary function and GPU queue submissions.
+class GraphicsCaptureEventProcessor : public orbit_capture_client::CaptureEventProcessor {
+ public:
+  void ProcessEvent(const orbit_grpc_protos::ClientCaptureEvent& event) override {
+    switch (event.event_case()) {
+      case orbit_grpc_protos::ClientCaptureEvent::kFunctionCall:
+        ProcessFunctionCall(event.function_call());
+        break;
+      case orbit_grpc_protos::ClientCaptureEvent::kGpuQueueSubmission:
+        ProcessGPUQueueSubmission(event.gpu_queue_submission());
+        break;
+      default:
+        break;
+    }
+  }
+
+  void ProcessGPUQueueSubmission(const orbit_grpc_protos::GpuQueueSubmission& submission) {
+    submissions_.emplace_back(submission);
+  }
+
+  void ProcessFunctionCall(const orbit_grpc_protos::FunctionCall& function_call) {
+    if (function_call.function_id() == kQueuePresentFunctionId) {
+      uint64_t start_timestamp_ns = function_call.end_timestamp_ns() - function_call.duration_ns();
+      frame_start_boundary_timestamps_.emplace_back(start_timestamp_ns);
+    }
+  }
+
+  // Calculate the GPU frame times in [ms] by measuring the amount of time spent by the GPU
+  // executing a GPU Job, from any queue that was submitted between two sequential vkQueuePresentKHR
+  // calls. The algorithm take into account that is possible that two jobs from different queue's
+  // can overlap, so the time is not account for multiple times.
+  //
+  // This function also generates:
+  // - The corresponding GPU frame time histogram, that is divided in 1024 buckets, each bucket X
+  // represents how many frames have a duration between [X, X+1) ms.
+  // - Average GPU frame times[ms]
+  void CalculateGPUTimes() {
+    LOG("Calculating GPU Times");
+    uint64_t total_duration_ns = 0;
+    size_t current_submission = 0;
+    gpu_time_distribution_.fill(0u);
+    for (uint64_t next_frame_start_timestamp : frame_start_boundary_timestamps_) {
+      uint64_t frame_time_ns = 0;
+      std::vector<std::pair</*begin*/ uint64_t, /*end*/ uint64_t>> command_buffer_timestamps;
+      while (current_submission < submissions_.size()) {
+        const orbit_grpc_protos::GpuQueueSubmission& submission = submissions_[current_submission];
+        if (submission.meta_info().pre_submission_cpu_timestamp() >= next_frame_start_timestamp) {
```

Comment:
@florian-kuebler ? Otherwise it might be worthwhile to sort no matter what.

Reply:
For me the question arose partly because you are not sorting here, but you are sorting below, so as a reader I have to assume it's not necessary here but necessary below, but it's not obvious why. 

Considering from the perspective of someone looking at this code at a later time (maybe because of a bug, etc), I would make assumptions about sort order explicit and comment why are you (not) sorting, in particular, because these are external assumptions about a different component (the layer supplying the data). 

As examples, if you decide to not sort:
```
// We assume timestamps come in order, so we do not sort them. 
[...]
```

If you decide to sort:
```
// Timestamps should come in order, but we sort regardless to be careful and since runtime 
// doesn't matter that much here.
sort(...)
```

These comments make it obvious why you wrote the code the way you did and make it easy to change it if assumptions change (or we find out they are incorrect).

Reply:
So the events are actually coming in order by CPU timestamp. But there is a [bit work done on the layer](https://github.com/google/orbit/blob/main/src/OrbitVulkanLayer/SubmissionTracker.h#L486), so it not obvious here.

I suggest to either add a comment or even just perform the sort here once as suggested.

Reply:
The reason I'm sorting the command buffer based on GPU times is because in Vulkan they can be executed out of order and also command buffers from different queues are able to run in parallel.

For the CPU timestamps the assumption is that they always be received in order

Reply:
Thanks for the explanation. I suggest to add a comment here, that you assume the sorting (or just sort for consistency), and also add a comment below, that you need to sort based on what you said.

Reply:
Added the comment so its clear why it needs sorting